src/sessions/VirtualPageLoadTimer.ts

```diff
@@ -0,0 +1,268 @@
+import { NavigationEvent } from '../events/navigation-event';
+import { PERFORMANCE_NAVIGATION_EVENT_TYPE } from '../plugins/utils/constant';
+import { MonkeyPatch, MonkeyPatched } from '../plugins/MonkeyPatched';
+import { Config } from '../orchestration/Orchestration';
+import { RecordEvent } from '../plugins/Plugin';
+import { PageManager, Page } from '../sessions/PageManager';
+
+type Fetch = typeof fetch;
+type Send = () => void;
+
+/**
+ * Maintains the core logic for virtual page load timing functionality.
+ * (1) Holds all virtual page load timing related resources
+ * (2) Intercepts outgoing XMLHttpRequests and Fetch requests and listens for DOM changes
+ * (3) Records virtual page load
+ */
+export class VirtualPageLoadTimer extends MonkeyPatched {
+    /** Latest interaction time by user on the document */
+    public latestInteractionTime: number;
+    /** Unique ID of virtual page load periodic checker. */
+    private periodicCheckerId;
+    /** Unique ID of virtual page load timeout checker. */
+    private timeoutCheckerId;
+    /** Observer to liten for DOM changes. */
+    private domMutationObserver: MutationObserver;
+    /** Set to hold outgoing XMLHttpRequests for current virtual page. */
+    private ongoingRequests: Set<XMLHttpRequest>;
+    /** Buffer to hold outgoing XMLHttpRequests before new page is created. */
+    private requestBuffer: Set<XMLHttpRequest>;
+    /** Indicate the number of active Fetch requests. */
+    private fetchCounter: number;
+    /** Indicate the status of the current Page */
+    private isPageLoaded: boolean;
+    /** Indicate the current page's load end time value. */
+    private latestEndTime: number;
+
+    private config: Config;
+    private pageManager: PageManager;
+    private record: RecordEvent;
+
+    constructor(pageManager: PageManager, config: Config, record: RecordEvent) {
+        super();
+        this.periodicCheckerId = undefined;
+        this.timeoutCheckerId = undefined;
+        this.domMutationObserver = new MutationObserver(this.resetInterval);
+        this.ongoingRequests = new Set<XMLHttpRequest>();
+        this.requestBuffer = new Set();
+        this.fetchCounter = 0;
+        this.isPageLoaded = true;
+        this.latestEndTime = 0;
+        this.latestInteractionTime = 0;
+
+        this.config = config;
+        this.pageManager = pageManager;
+        this.record = record;
+        this.enable();
+
+        // Start tracking the timestamps
+        document.body.addEventListener(
+            'mousedown',
+            this.updateLatestInteractionTime
+        );
+        document.body.addEventListener(
+            'keydown',
+            this.updateLatestInteractionTime
+        );
+        document.body.addEventListener(
+            'mouseleave',
+            this.updateLatestInteractionTime
+        );
```

Comment:
question: What is the reason for including this? Is there a risk that this could overwrite the actual interaction time?

Reply:
Synced offline, remove `eventListener` on `mouseleave`